Add blsr

Wraith2 opened the pull request:
closes https://github.com/dotnet/runtime/issues/63286

This PR adds a new `LowerAndOpToResetLowestSetBit` in xarch to recognise the `AND(x, SUB(x, 1)` pattern and lower it to a `blsr` where it is likely to be profitable. 

## benchmarks.run.windows.x64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 13363029 (overridden on cmd)
Total bytes of diff: 13363021 (overridden on cmd)
Total bytes of delta: -8 (-0.00 % of base)
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
          -2 : 12409.dasm (-0.31% of base)
          -2 : 15708.dasm (-0.99% of base)
          -2 : 12421.dasm (-0.03% of base)
          -2 : 15719.dasm (-0.26% of base)

4 total files with Code Size differences (4 improved, 0 regressed), 3 unchanged.

Top method improvements (bytes):
          -2 (-0.31% of base) : 12409.dasm - System.Diagnostics.Tracing.ManifestBuilder:AddKeyword(System.String,long):this
          -2 (-0.03% of base) : 12421.dasm - System.Diagnostics.Tracing.ManifestBuilder:CreateManifestString():System.String:this
          -2 (-0.26% of base) : 15719.dasm - System.IO.RandomAccess:TryPrepareScatterGatherBuffers(System.Collections.Generic.IReadOnlyList`1[Memory`1],MemoryHandler,byref,byref,byref):bool
          -2 (-0.99% of base) : 15708.dasm - System.Runtime.InteropServices.NativeMemory:AlignedAlloc(long,long):long

Top method improvements (percentages):
          -2 (-0.99% of base) : 15708.dasm - System.Runtime.InteropServices.NativeMemory:AlignedAlloc(long,long):long
          -2 (-0.31% of base) : 12409.dasm - System.Diagnostics.Tracing.ManifestBuilder:AddKeyword(System.String,long):this
          -2 (-0.26% of base) : 15719.dasm - System.IO.RandomAccess:TryPrepareScatterGatherBuffers(System.Collections.Generic.IReadOnlyList`1[Memory`1],MemoryHandler,byref,byref,byref):bool
          -2 (-0.03% of base) : 12421.dasm - System.Diagnostics.Tracing.ManifestBuilder:CreateManifestString():System.String:this

4 total methods with Code Size differences (4 improved, 0 regressed), 3 unchanged.

```

</details>

--------------------------------------------------------------------------------

## coreclr_tests.pmi.windows.x64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 127528257 (overridden on cmd)
Total bytes of diff: 127528255 (overridden on cmd)
Total bytes of delta: -2 (-0.00 % of base)
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
          -2 : 230250.dasm (-1.41% of base)

1 total files with Code Size differences (1 improved, 0 regressed), 8 unchanged.

Top method improvements (bytes):
          -2 (-1.41% of base) : 230250.dasm - V8.Crypto.BigInteger:bitCount():int:this

Top method improvements (percentages):
          -2 (-1.41% of base) : 230250.dasm - V8.Crypto.BigInteger:bitCount():int:this

1 total methods with Code Size differences (1 improved, 0 regressed), 8 unchanged.

```

</details>

--------------------------------------------------------------------------------

## libraries.pmi.windows.x64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 46215753 (overridden on cmd)
Total bytes of diff: 46215748 (overridden on cmd)
Total bytes of delta: -5 (-0.00 % of base)
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
          -3 : 142808.dasm (-0.28% of base)
          -2 : 215581.dasm (-1.47% of base)

2 total files with Code Size differences (2 improved, 0 regressed), 20 unchanged.

Top method improvements (bytes):
          -3 (-0.28% of base) : 142808.dasm - System.Xml.Xsl.IlGen.XmlILVisitor:MatchesNodeKinds(System.Xml.Xsl.Qil.QilTargetType,System.Xml.Xsl.XmlQueryType,System.Xml.Xsl.XmlQueryType):bool:this
          -2 (-1.47% of base) : 215581.dasm - System.Numerics.BigInteger:get_IsPowerOfTwo():bool:this

Top method improvements (percentages):
          -2 (-1.47% of base) : 215581.dasm - System.Numerics.BigInteger:get_IsPowerOfTwo():bool:this
          -3 (-0.28% of base) : 142808.dasm - System.Xml.Xsl.IlGen.XmlILVisitor:MatchesNodeKinds(System.Xml.Xsl.Qil.QilTargetType,System.Xml.Xsl.XmlQueryType,System.Xml.Xsl.XmlQueryType):bool:this

2 total methods with Code Size differences (2 improved, 0 regressed), 20 unchanged.

```

</details>

--------------------------------------------------------------------------------

## libraries_tests.pmi.windows.x64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 118063813 (overridden on cmd)
Total bytes of diff: 118063785 (overridden on cmd)
Total bytes of delta: -28 (-0.00 % of base)
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
          -2 : 111078.dasm (-0.59% of base)
          -2 : 110781.dasm (-9.09% of base)
          -2 : 111083.dasm (-0.59% of base)
          -2 : 110793.dasm (-9.09% of base)
          -2 : 6689.dasm (-9.09% of base)
          -2 : 326883.dasm (-0.58% of base)
          -2 : 187034.dasm (-1.27% of base)
          -2 : 326891.dasm (-0.39% of base)
          -2 : 187035.dasm (-1.27% of base)
          -2 : 187036.dasm (-1.24% of base)
          -2 : 187037.dasm (-1.27% of base)
          -2 : 187038.dasm (-1.25% of base)
          -2 : 324730.dasm (-8.70% of base)
          -2 : 187039.dasm (-1.27% of base)

14 total files with Code Size differences (14 improved, 0 regressed), 10 unchanged.

Top method improvements (bytes):
          -2 (-9.09% of base) : 110781.dasm - enumOfType@48-38:Invoke(long):bool:this
          -2 (-9.09% of base) : 110793.dasm - enumOfType@50-44:Invoke(long):bool:this
          -2 (-9.09% of base) : 6689.dasm - Microsoft.CodeAnalysis.Shared.Utilities.IntegerUtilities:CountOfBitsSet(long):int
          -2 (-0.59% of base) : 111078.dasm - reflectShrinkObj@256-17:GenerateNext(byref):int:this
          -2 (-0.59% of base) : 111083.dasm - reflectShrinkObj@257-18:GenerateNext(byref):int:this
          -2 (-1.27% of base) : 187034.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_long(long,bool)
          -2 (-1.24% of base) : 187036.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nint_32(int,bool)
          -2 (-1.27% of base) : 187037.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nint_64(long,bool)
          -2 (-1.25% of base) : 187038.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nuint_32(int,bool)
          -2 (-1.27% of base) : 187039.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nuint_64(long,bool)
          -2 (-1.27% of base) : 187035.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_ulong(long,bool)
          -2 (-0.58% of base) : 326883.dasm - System.Runtime.InteropServices.Tests.NativeMemoryTests:AlignedAllocTest(int):this
          -2 (-0.39% of base) : 326891.dasm - System.Runtime.InteropServices.Tests.NativeMemoryTests:AlignedReallocTest(int):this
          -2 (-8.70% of base) : 324730.dasm - System.Tests.BinaryNumberHelper`1[Int64][System.Int64]:IsPow2(long):bool

Top method improvements (percentages):
          -2 (-9.09% of base) : 110781.dasm - enumOfType@48-38:Invoke(long):bool:this
          -2 (-9.09% of base) : 110793.dasm - enumOfType@50-44:Invoke(long):bool:this
          -2 (-9.09% of base) : 6689.dasm - Microsoft.CodeAnalysis.Shared.Utilities.IntegerUtilities:CountOfBitsSet(long):int
          -2 (-8.70% of base) : 324730.dasm - System.Tests.BinaryNumberHelper`1[Int64][System.Int64]:IsPow2(long):bool
          -2 (-1.27% of base) : 187034.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_long(long,bool)
          -2 (-1.27% of base) : 187037.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nint_64(long,bool)
          -2 (-1.27% of base) : 187039.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nuint_64(long,bool)
          -2 (-1.27% of base) : 187035.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_ulong(long,bool)
          -2 (-1.25% of base) : 187038.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nuint_32(int,bool)
          -2 (-1.24% of base) : 187036.dasm - System.Numerics.Tests.BitOperationsTests:BitOps_IsPow2_nint_32(int,bool)
          -2 (-0.59% of base) : 111078.dasm - reflectShrinkObj@256-17:GenerateNext(byref):int:this
          -2 (-0.59% of base) : 111083.dasm - reflectShrinkObj@257-18:GenerateNext(byref):int:this
          -2 (-0.58% of base) : 326883.dasm - System.Runtime.InteropServices.Tests.NativeMemoryTests:AlignedAllocTest(int):this
          -2 (-0.39% of base) : 326891.dasm - System.Runtime.InteropServices.Tests.NativeMemoryTests:AlignedReallocTest(int):this

14 total methods with Code Size differences (14 improved, 0 regressed), 10 unchanged.

```

</details>

Improvement example from `V8.Crypto.BigInteger:bitCount():int`

<details>

<summary>improvement diff</summary>

```diff
@@ -14,7 +14,7 @@
 ;  V04 OutArgs      [V04    ] (  1,  1   )  lclBlk (32) [rsp+00H]   "OutgoingArgSpace"
 ;  V05 tmp1         [V05,T03] (  3, 24   )     ref  ->   r8         class-hnd "Inlining Arg"
 ;  V06 tmp2         [V06,T01] (  4, 40   )     int  ->   r8         "Inline stloc first use temp"
-;  V07 tmp3         [V07,T00] (  6,144   )     int  ->  r10         "Inlining Arg"
+;  V07 tmp3         [V07,T00] (  5,112   )     int  ->  r10         "Inlining Arg"
 ;  V08 tmp4         [V08,T04] (  3, 24   )     ref  ->   r8         "arr expr"
 ;  V09 cse0         [V09,T08] (  2,  5   )     ref  ->  rax         "CSE - aggressive"
 ;  V10 cse1         [V10,T07] (  3,  6   )     int  ->  rcx         "CSE - aggressive"
@@ -63,12 +63,11 @@ G_M16143_IG03:        ; gcrefRegs=00000001 {rax}, byrefRegs=00000000 {}, byref,
        align    [4 bytes for IG04]
 						;; bbWeight=4    PerfScore 62.00
 G_M16143_IG04:        ; gcrefRegs=00000001 {rax}, byrefRegs=00000000 {}, loop=IG04, byref, isz
-       lea      r9d, [r10-1]
-       and      r10d, r9d
+       blsr     r10d, r10d
        inc      r8d
        test     r10d, r10d
        jne      SHORT G_M16143_IG04
-						;; bbWeight=16    PerfScore 36.00
+						;; bbWeight=16    PerfScore 32.00
 G_M16143_IG05:        ; gcrefRegs=00000001 {rax}, byrefRegs=00000000 {}, byref, isz
        add      edi, r8d
        inc      edx
@@ -97,7 +96,7 @@ G_M16143_IG09:        ; gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref
        int3     
 						;; bbWeight=0    PerfScore 0.00
 
-; Total bytes of code 142, prolog size 10, PerfScore 136.95, instruction count 48, allocated bytes for code 142 (MethodHash=2afdc0f0) for method V8.Crypto.BigInteger:bitCount():int:this
+; Total bytes of code 140, prolog size 10, PerfScore 132.75, instruction count 47, allocated bytes for code 140 (MethodHash=2afdc0f0) for method V8.Crypto.BigInteger:bitCount():int:this
 ; ============================================================
 
 Unwind Info:
```

</details>

All improvements. no regressions.

this was created as a first attempt to contribute to the jit with extensive help from @EgorBo @SingleAccretion and others from #lowlevel on the c# discord server. I can change whatever you need me to as long as you can explain what you need to me clearly. If this is a success i can continue to do the same for some of the other simple BMI1 instruction patterns.

/cc @dotnet/jit-contrib

Wraith2 committed:
add lowering to blsr instruction

Wraith2 committed:
fixup next for replacement node and clear containment

Wraith2 committed:
add check for user being boolean to prevent optimization of LEA, TEST case.

Wraith2 committed:
add comment and rename variables

Wraith2 committed:
ifdef optimization on HWIntrinsics feature and fix formatting

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+        GenTree* addOp1 = op2->gtGetOp1();
+        GenTree* addOp2 = op2->gtGetOp2();
+        if (addOp2->IsIntegralConst(-1) && addOp1->OperIs(GT_LCL_VAR) &&
+            addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum() &&
```
```suggestion
            (addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum()) &&
```
See [the coding conventions](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#111-logical-and-arithmetic-expressions).

Wraith2 replied:
Done.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+        GenTree* addOp2 = op2->gtGetOp2();
+        if (addOp2->IsIntegralConst(-1) && addOp1->OperIs(GT_LCL_VAR) &&
+            addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum() &&
+            ((op1->TypeGet() == TYP_LONG && comp->compOpportunisticallyDependsOn(InstructionSet_BMI1_X64)) ||
```
```suggestion
            ((op1->TypeIs(TYP_LONG) && comp->compOpportunisticallyDependsOn(InstructionSet_BMI1_X64)) ||
```
(Avoids the need to parenthesize `op1->TypeGet() == TYP_LONG`)

Wraith2 replied:
Changed.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                // if we prevent the use of BSLR here then the OptimizeConstCompare lowering may generate
+                //   TEST_NE or TEST_EQ leading to a cheaper "lea, test"
+                GenTree* user = use.User();
+                if (user != nullptr && !user->OperIs(GT_EQ, GT_NE))
```
```suggestion
                if (!user->OperIs(GT_EQ, GT_NE))
```
If `TryGetUse` returns `true`, `user` is never `nullptr`.

Wraith2 replied:
It'd be nice if things like this were in the function header comments. Overall more comments and explanation would make this code much easier to work with.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                GenTree* user = use.User();
+                if (user != nullptr && !user->OperIs(GT_EQ, GT_NE))
+                {
+                    GenTreeHWIntrinsic* replacementNode =
```
Nit: `replacementNode` -> `blsrNode`.

Wraith2 replied:
Done.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                {
+                    GenTreeHWIntrinsic* replacementNode =
+                        comp->gtNewScalarHWIntrinsicNode(op1->TypeGet(), op1,
+                                                         op1->TypeGet() == TYP_INT
+                                                             ? NamedIntrinsic::NI_BMI1_ResetLowestSetBit
+                                                             : NamedIntrinsic::NI_BMI1_X64_ResetLowestSetBit);
```
This should use `node`s types, both for the new node and the intrinsics. The reason is that for `AND` we're guaranteed to only ever see `TYP_INT/TYP_LONG` (with one exception that is not relevant here), while `op1` can be of a small type (`TYP_UBYTE` and friends).

Nit: `TypeGet()` -> `TypeIs`

Wraith2 replied:
All done.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+
+                    use.ReplaceWith(replacementNode);
+
+                    op1->ClearContained();
```
```suggestion
```
No need to `ClearContained`, as we haven't yet established the containment for `node`.

Wraith2 replied:
Removed.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                    return replacementNode;
+                }
+            }
+            else
+            {
+                node->SetUnusedValue();
+            }
```
```suggestion
```
This can be omitted - `IsUnuseValue` will already be set on `node` in that case.

Wraith2 replied:
Removed.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                node->SetUnusedValue();
+            }
+        }
+    }
+    return nullptr;
```
```suggestion
    }

    return nullptr;
```

Wraith2 replied:
Done.

Wraith2 commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+        if (addOp2->IsIntegralConst(-1) && addOp1->OperIs(GT_LCL_VAR) &&
+            addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum() &&
+            ((op1->TypeGet() == TYP_LONG && comp->compOpportunisticallyDependsOn(InstructionSet_BMI1_X64)) ||
+             comp->compOpportunisticallyDependsOn(InstructionSet_BMI1)) // if op1->TypeGet()!=TYP_LONG then it must be
+                                                                        // TYP_INT
+            )
```
Not forcing the weird format with jit-tidy would make it less weird as well. There are many better ways to layout this code than it wants, this is not a good format it is only a compact one.

Wraith2 replied:
Done.

Wraith2 commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
         LowerNode(cast);
     }
 }
+
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerHWIntrinsicToScalar: Lowers a tree AND(X, ADD(X, -1) to HWIntrinsic::ResetLowestSetBit
+//     Returns the recplacement node if one is created else nullptr indicating no replacement
+//
+// Parameters
+//     tree   -   GentreePtr of binOp
+//
```
I copied my comment from one of the other nearby functions in the file. Perhaps the rest of the file should be updated to follow those same conventions.

SingleAccretion replied:
It is very common practice to update function headers of code being refactored / added. It is expected that the function headers will be inconsistent until all are converted to the "modern" format.

Wraith2 replied:
Updated. I think the current state is silly. If it needs to conform to conventions in which case it should be converted not left in a confused state.

Wraith2 commented on src/coreclr/jit/lower.cpp:
```diff
@@ -5133,41 +5133,9 @@ GenTree* Lowering::LowerAdd(GenTreeOp* node)
-GenTree* Lowering::LowerBinaryArithmetic(GenTreeOp* node)
+GenTree* Lowering::LowerBinaryArithmeticCommon(GenTreeOp* node)
 {
-    // TODO-CQ-XArch: support BMI2 "andn" in codegen and condition
-    // this logic on the support for the instruction set on XArch.
-    CLANG_FORMAT_COMMENT_ANCHOR;
```
The current implementation was intended to be arm specific so I moved it there. If someone implement xarch andn and finds the code is identical it can be moved up to the common location but for now it feels cleaner to have something that's arm specific in the arm specific file than it does to have if ifdefed for arm in the common file.

SingleAccretion replied:
> The current implementation was intended to be arm specific so I moved it there

I wrote that code so can say with confidence it was very intentional that it lives in the common function :).

If we are to separate them however, as it is true this current implementation is ARMARCH-only, `LowerBinaryArithmeticCommon` needs to be renamed back to `LowerBinaryArithmetic` and deleted (as it now does not do anything but call the arch-specific function). Also the `TODO-CQ-XArch: support BMI2 "andn"..."` needs to mention this ARMARCH logic.

Wraith2 replied:
If I move it back up to common I have the problem that I need to know if the node has been replaced and containment checked by the hwintrinsic function for that purpose or I need have arm one path on arm and one path on xarch which doesn't make any sense to do in the common file when there's a file for each one of those things. I suppose this is why people seem to change nodes not replace them.

Wraith2 replied:
Dealt with.

Wraith2 commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -159,6 +159,40 @@ GenTree* Lowering::LowerMul(GenTreeOp* mul)
+    // this logic on the support for the instruction set on XArch.
+    CLANG_FORMAT_COMMENT_ANCHOR;
+
+    GenTree* replacement = nullptr;
```
Done.

Wraith2 commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+//     tree   -   GentreePtr of binOp
+//
+GenTree* Lowering::LowerAndOpToResetLowestSetBit(GenTree* node)
+{
```
Done.

Wraith2 commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+            if (BlockRange().TryGetUse(node, &use))
+            {
+                // If we allow the use of BSLR where the parent is NE or EQ it will generate "blsr, test"
+                // if we prevent the use of BSLR here then the OptimizeConstCompare lowering may generate
```
Changed.

Wraith2 committed:
address reviewer feedback

Wraith2 committed:
move arm ANDN back up to common

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3720,75 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
     }
 }
+
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerAndOpToResetLowestSetBit: Lowers a tree AND(X, ADD(X, -1) to HWIntrinsic::ResetLowestSetBit
+//
+// Arguments:
+//    andNode - GentreePtr of binOp
+//
+// Return Value:
+//    Returns the replacement node if one is created else nullptr indicating no replacement
+//
+// Assumptions:
+//    andNode is GT_AND and (TYP_INT or TYP_LONG)
+//
```
```suggestion
//----------------------------------------------------------------------------------------------
// Lowering::LowerAndOpToResetLowestSetBit: Optimize AND(X, ADD(X, -1) to HWIntrinsic::ResetLowestSetBit
//
// Arguments:
//    andNode - GT_AND node of integral type
//
// Return Value:
//    Returns the next node to lower if the optimization was successful, "nullptr" otherwise.
//
```
(I would consider changing the name from `Lower` to `Optimize` too)

Wraith2 replied:
I've removed assumptions. Though it would make sense to me that if we're asserting something at the beginning or end of the method that those are sensible things to document as pre and post conditions.

SingleAccretion replied:
I agree, and I didn't mean to drop that part, just move it to the argument description: (right now it still reads `GentreePtr of binOp`, where `GentreePtr` has been [not a thing](https://github.com/dotnet/coreclr/pull/16027) for quite some time).

Wraith2 replied:
Fixed.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3720,75 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+        if (addOp2->IsIntegralConst(-1) && addOp1->OperIs(GT_LCL_VAR) &&
+            (addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum()) &&
+            ((op1->TypeIs(TYP_LONG) && comp->compOpportunisticallyDependsOn(InstructionSet_BMI1_X64)) ||
+             comp->compOpportunisticallyDependsOn(InstructionSet_BMI1))
+            )
```
I would expect you can combine the `)` with the `comp->compOpportunisticallyDependsOn(InstructionSet_BMI1))` now.

Wraith2 replied:
The coding conventions say:

> There should be parentheses around all unary and binary expressions if they are contained within other expressions. We prefer to over-specify parentheses instead of requiring developers to memorize the complete C++ precedence rules. This is especially true for && and ||, whose precedence relationship is often forgotten.

which seems to be ok with the additional parents around the `((islong && issupportedlong) || issupportedint)`

SingleAccretion replied:
Yes, I didn't mean to drop it, but move it up a line, i. e.:
```
              comp->compOpportunisticallyDependsOn(InstructionSet_BMI1))
            ) 
```
->
```
              comp->compOpportunisticallyDependsOn(InstructionSet_BMI1))) 
```

Wraith2 replied:
I think the formatter has moved it up.

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3720,75 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                    BlockRange().Remove(op2);
+                    BlockRange().Remove(addOp1);
+                    BlockRange().Remove(addOp2);
+                    JITDUMP("Remove [%06u], [%06u]\n", andNode->gtTreeID, andNode->gtTreeID);
```
This `JITDUMP` is not terribly correct (prints the same node twice, and doesn't print the rest of the removed nodes). I think it could be removed, or if not, it should print IDs of all the removed nodes.

(It would be nice to print the "Removing unused node" message and each of them separately with `DISPNODE`, but it's not worth the code)
(We should probably add an option to `LIR::Range::Remove` to print this message, activated with a parameter, but that's unrelated cleanup)

Wraith2 replied:
Removed. We Jitdump should provide useful information. Again this is a case where I looked at an existing function and did the same thing though so code quality elsewhere doesn't seem to meet the current standard.

SingleAccretion commented on src/coreclr/jit/lower.h:
```diff
@@ -297,7 +297,8 @@ class Lowering final : public Phase
     GenTree* LowerAdd(GenTreeOp* node);
     GenTree* LowerMul(GenTreeOp* mul);
-    GenTree* LowerBinaryArithmetic(GenTreeOp* node);
+    GenTree* LowerBinaryArithmeticCommon(GenTreeOp* node);
```
```suggestion
    GenTree* LowerBinaryArithmeticCommon(GenTreeOp* binOp);
```

Wraith2 replied:
Done.

Wraith2 replied:
Done.

SingleAccretion commented on src/coreclr/jit/lower.h:
```diff
@@ -343,6 +344,7 @@ class Lowering final : public Phase
     void LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node);
     void LowerHWIntrinsicGetElement(GenTreeHWIntrinsic* node);
     void LowerHWIntrinsicWithElement(GenTreeHWIntrinsic* node);
+    GenTree* LowerAndOpToResetLowestSetBit(GenTree* node);
```
```suggestion
    GenTree* LowerAndOpToResetLowestSetBit(GenTree* andNode);
```
Also, this should be a `GenTreeOp*`.

Wraith2 replied:
Done and done.

Wraith2 committed:
more review feedback

Wraith2 committed:
feedback

SingleAccretion commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3720,70 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                // if we prevent the use of BSLR here then "OptimizeConstCompare" may generate TEST_NE or
+                //   TEST_EQ leading to a cheaper "lea, test"
+                GenTree* user = use.User();
+                if (!user->OperIs(GT_EQ, GT_NE))
```
This should also check that `user`'s second operand is constant (since we are anticipating `Optimize<Const>Compare`'s optimization).

Wraith2 replied:
Done.

Wraith2 committed:
add check for second user op being constant

EgorBo commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3720,70 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+                                                             ? NamedIntrinsic::NI_BMI1_ResetLowestSetBit
+                                                             : NamedIntrinsic::NI_BMI1_X64_ResetLowestSetBit);
+
+                    JITDUMP("Lower: optimize AND(X, ADD(X, -1)): ");
```
have you checked the DOTNET_JITDUMP with these logs, you might need to add \n inside JITDUMP and a `JITDUMP("\n") after `DISPNODE(blsrNode);`

Wraith2 replied:
There are no uses of a newline only jitdump after a dispnode so i don't think that's needed. I will add a newline at the end of the jitdumps.

EgorBo commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -3693,6 +3720,70 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
+            if (BlockRange().TryGetUse(andNode, &use))
+            {
+                // If we allow the use of BSLR where the parent is NE or EQ it will generate "blsr, test"
+                // if we prevent the use of BSLR here then "OptimizeConstCompare" may generate TEST_NE or
```
I don't get this part, C++ seems to be emitting blsr (a typo in the comment btw) in all cases: https://godbolt.org/z/fo8qGTe1q

I guess you mean the fact JIT doesn't realize blsr sets flags?

Wraith2 replied:
When I showed the results to @SingleAccretion he suggested adding the code to check the use so that it would avoid the pattern. He'll have to explain the reasoning behind that.

SingleAccretion replied:
> I guess you mean the fact JIT doesn't realize blsr sets flags?

Yes. I suppose fixing it would require recognizing `blsr` here:

https://github.com/dotnet/runtime/blob/5fdc58d40141cc48d614a3e930f83f12cd538dad/src/coreclr/jit/lower.cpp#L2698-L2714

Wraith2 replied:
I think that would just be adding the check `|| op1.OperIs(GT_HWINSTRINSIC) && (op1->AsHWIntrinsic()->GetHWIntrinsicId() == NamedIntrinsic::NI_BMI1_ResetLowestSetBit || op1->AsHWIntrinsic()->GetHWIntrinsicId() == NamedIntrinsic::NI_BMI1_X64_ResetLowestSetBit)`. Are you asking me to do this and remove the use check in the method I've added or is this just discussion?

SingleAccretion replied:
> Are you asking me to do this and remove the use check in the method I've added

I think it would be a good idea to do it as part of this change, yes.

Wraith2 replied:
Done.
I had to change the instruction definition to include the flags so that they could later be checked. The diff is now all improvements, no regressions.

Wraith2 committed:
add jitdump newlines

Wraith2 committed:
remove lower equality check and set blsr instruction flags so that they can be detected and avoided.

Wraith2 committed:
add parens and tidy

EgorBo commented on src/coreclr/jit/lower.cpp:
```diff
@@ -2708,10 +2708,16 @@ GenTree* Lowering::OptimizeConstCompare(GenTree* cmp)
+            (op1->OperIs(GT_AND, GT_OR, GT_XOR, GT_ADD, GT_SUB, GT_NEG)
+#ifdef FEATURE_HW_INTRINSICS
+             || emitter::DoesWriteZeroFlag(
+                    HWIntrinsicInfo::lookupIns(op1->AsHWIntrinsic()->GetHWIntrinsicId(), op1->TypeGet()))
```
here you don't check op1 for being GT_HWINTRINSIC before you call ->AsHWIntrinsic()

Wraith2 committed:
correct missing IsOper(GT_HWINTRINSIC) check

Wraith2 committed:
jit-format

Wraith2 commented:
I think all the CI failures are down to helix uploads not the PR itself. So, ready for more review changes.

SingleAccretion commented on src/coreclr/jit/lower.cpp:
```diff
@@ -2708,10 +2708,17 @@ GenTree* Lowering::OptimizeConstCompare(GenTree* cmp)
+#ifdef FEATURE_HW_INTRINSICS
+             || (op1->OperIs(GT_HWINTRINSIC) &&
+                 emitter::DoesWriteZeroFlag(
+                     HWIntrinsicInfo::lookupIns(op1->AsHWIntrinsic()->GetHWIntrinsicId(), op1->TypeGet())))
```
@tannergooding could you please take a look at this when you have time?

My concern is that it appears the second operand to `lookupIns` is "sometimes" "the base type" (presumably for SIMD intrinsics) and "sometimes" "the target type" (presumably for scalar intrinsics), and we're always passing "the target type" here (the node type). This feels prone to picking up wrong instructions, and I am not sure of the correct guard against it.

tannergooding replied:
Rather than taking the instruction, have an overload that takes the intrinsic node. The method can then use the node and do the relevant switch of "for scalar intrinsics, there is no `simdBaseType` so use the node type; for simd intrinsics use the simdBaseType"

Wraith2 replied:
Done. Added a lookupIns overload that takes the intrinsic node looks up the category and for scalar uses the node type otherwise gets the base simd type and then looks up the instruction.

Wraith2 committed:
reformatting as suggested by EgorBo

Wraith2 committed:
clang formatting is stupid

EgorBo commented:
build failures related

Wraith2 committed:
fix ordering of and check and add lookupIns(HWIntrinsic) overload

Wraith2 commented:
Green!

EgorBo reviewed (approved):
LGTM, Thanks! @dotnet/jit-contrib any thoughts? the diff is small but it improves https://github.com/dotnet/runtime/pull/63285 and potentially a new API in BitOperations

Wraith2 commented:
I think it might be a good idea to rename the method in introduced from `LowerAndOpToResetLowestSetBit` to `TryLowerAndOpToResetLowestSetBit` because in most cases it isn't going lower it, try feels more descriptive. That isn't a naming pattern that I've seen in the rest of the code though.

Wraith2 committed:
rename to prefix Try and remove unscoped enum prefixes

Wraith2 commented:
Updated to the Try variant of the name. Removed unscoped enum prefixes per request. An arm64 test seems to be failing but i don't think that's related.

Wraith2 commented:
@jakobbotsch, @EgorBo mentioned that you had concerns that checking for local var exposure was not stringent enough. Is there a way you'd prefer this to be handled?

jakobbotsch commented:
> @jakobbotsch, @EgorBo mentioned that you had concerns that checking for local var exposure was not stringent enough. Is there a way you'd prefer this to be handled?

I had a wondering whether we can see intermediate stores to the local, but looks like the actual position of those locals in LIR does not matter today. Sort of weird but if that is the case then I don't think anything more stringent is necessary.

EgorBo commented:
@Wraith2 Thanks!

Wraith2 commented:
yay!
